doc/aaa/ldap.md

```diff
@@ -0,0 +1,2617 @@
+
+
+; Logon Authorization from LDAP
+; Use this server for login authorization only
+sonic(config)# ldap-server 10.59.1.9 use-type pam
+
+sonic(config)# ldap-server pam-groupdn "cn=t1.loc.example.com,ou=hostaccess,dc=example,dc=com"
+sonic(config)# ldap-server pam-member-attribute "uniqueMember"
+sonic(config)# aaa authorization login ldap
+
+; sudo-ldap authorization
+sonic(config)# ldap-server sudoers-base "dc=example,dc=com"
+sonic(config)# aaa name-service sudoers ldap
+
+; Authentication from TACACS+
+sonic(config)# tacacs-server 10.59.1.15
+sonic(config)# aaa authentication failthrough enable
+sonic(config)# aaa authentication login-method tacacs+ local
+
+```
+
+# Flow Diagrams
+
+Please see the diagrams in section [Design](#design)
+
+# Error Handling
+
+PAM and NSS modules return errors as per [Linux-PAM](#Linux-PAM), and
+[NSS](#NSS-Reference) respectively.
+
+# Serviceability and Debug
+
+The LDAP PAM and NSS modules can be debugged using the "debug" option
+of AAA. (i.e. config aaa debug enable)
+
+# Warm Boot Support
+
+N/A
+
+# Scalability
+
+A maximum of 8 LDAP servers can be configured.
+
+# Unit Test
+
+## Prerequisites
+
+```
+  LDAP Server with User, Group information.
+  TACACS+/RADIUS Server with User passwords configure.
+```
+
+### LDAP data
+
+For sudo authorization (*sudoRole* objectclass), the LDAP server schema
+needs to be updated from DUT file /usr/share/doc/sudo-ldap/schema.OpenLDAP.
+Please see next section for some tips.
+
+
+```
+
+$ ldapsearch -x -b "dc=sji,dc=broadcom,dc=net" 
+# extended LDIF
+#
+# LDAPv3
+# base <dc=sji,dc=broadcom,dc=net> with scope subtree
+# filter: (objectclass=*)
+# requesting: ALL
+#
+
+# sji.broadcom.net
+dn: dc=sji,dc=broadcom,dc=net
+objectClass: top
+objectClass: dcObject
+objectClass: organization
+o: sji.broadcom.net
+dc: sji
+
+# admin, sji.broadcom.net
+dn: cn=admin,dc=sji,dc=broadcom,dc=net
+objectClass: simpleSecurityObject
+objectClass: organizationalRole
+cn: admin
+description: LDAP administrator
+
+# Group, sji.broadcom.net
+dn: ou=Group,dc=sji,dc=broadcom,dc=net
+objectClass: organizationalUnit
+ou: Group
+
+# People, sji.broadcom.net
+dn: ou=People,dc=sji,dc=broadcom,dc=net
+objectClass: organizationalUnit
+ou: People
+
+# Sudoers, sji.broadcom.net
+dn: ou=Sudoers,dc=sji,dc=broadcom,dc=net
+objectClass: organizationalUnit
+ou: Sudoers
+
+# sudo, Group, sji.broadcom.net
+dn: cn=sudo,ou=Group,dc=sji,dc=broadcom,dc=net
+objectClass: top
+objectClass: posixGroup
+gidNumber: 27
+memberUid: arun
+memberUid: ldapuser
+cn: sudo
+
+# docker, Group, sji.broadcom.net
+dn: cn=docker,ou=Group,dc=sji,dc=broadcom,dc=net
+objectClass: top
+objectClass: posixGroup
+gidNumber: 999
+memberUid: arun
+memberUid: ldapuser
+memberUid: raduser1
+cn: docker
+
+# arun, People, sji.broadcom.net
+dn: cn=arun,ou=People,dc=sji,dc=broadcom,dc=net
+objectClass: top
+objectClass: account
+objectClass: posixAccount
+objectClass: shadowAccount
+cn: arun
+uid: arun
+uidNumber: 1200
```

Comment:
Do we know what happens if there is a conflict between the local DB (`/etc/passwd`, `/etc/group`) and the LDAP server? For example, what if the local `/etc/passwd` assigns the uid 1000 to a user named "`john`", but the same uid is assigned to a user named "`bob`" on the LDAP server.